File: src/lib/components/model/MaintenancePart.js
export class MaintenancePart {
    /** @type {Number | null} id*/
    id
    /** @type {Number} kilometer*/
    kilometer
    /** @type {String | null} name*/
    name
    /** @type {Boolean | null} is_part_of_the_maintenance*/
    is_part_of_the_maintenance
    /** @type {String | null} reference*/
    reference
    /** @type {PartLocation | null} part_location*/
    part_location
    /** @type {String | null} oem_reference*/
    oem_reference
    /** @type {Boolean | null} made_by*/
    diy
    /** @type {String | null} made_by*/
    made_by
    /** @type {Number | null} next_maintenance*/
    next_maintenance


    constructor() {
        this.id = 0;
        this.kilometer = null;
        this.name = null;
        this.is_part_of_the_maintenance = false;
        this.reference = null;
        this.part_location = null;
        this.oem_reference = null;
        this.diy = false;
        this.made_by = null;
        this.next_maintenance = null;
    }

    is_valid() {
        return this.kilometer !== null &&
            this.name !== null &&
            this.reference !== null &&
            this.part_location !== null &&
            this.oem_reference !== null &&
            this.made_by !== null &&
            this.next_maintenance !== null
    }

}


/**
 * @readonly
 * @enum {String}
 */
const PartLocation = {
    EngineCompartment: 'EngineCompartment',
    ExteriorBodywork: 'ExteriorBodywork',
    Interior: 'Interior',
    Chassis: 'Chassis',
    Tires: 'Tires',
    ElectronicSystems: 'ElectronicSystems',
    TransmissionSystem: 'TransmissionSystem',
    SuspensionSystem: 'SuspensionSystem',
    BrakingSystem: 'BrakingSystem',
};

File: src-tauri/src/models/maintenance_part.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Clone, Debug, PartialEq, PartialOrd, sqlx::Type, Deserialize, Serialize)]
#[sqlx(type_name = "part_location", rename_all = "camelCase")]
pub enum PartLocation {
    EngineCompartment,
    ExteriorBodywork,
    Interior,
    Chassis,
    Tires,
    ElectronicSystems,
    TransmissionSystem,
    SuspensionSystem,
    BrakingSystem,
}

#[derive(Serialize, Deserialize, FromRow, Clone, Debug)]
pub struct MaintenancePartDb {
    pub id: i32,
    pub kilometer: i32,
    pub name: String,
    pub reference: String,
    pub part_location: PartLocation,
    pub oem_reference: String,
    pub diy: bool,
    pub made_by: Option<String>,
    pub is_part_of_the_maintenance: bool,
    pub next_maintenance: Option<i32>,
}
